be/src/exec/olap_scanner.cpp

```diff
@@ -150,6 +150,20 @@ Status OlapScanner::_init_params(
     _params.aggregation = _aggregation;
     _params.version = Version(0, _version);
 
+    // acquire tablet rowset readers at the beginning of the scan node 
+    // to prevent this case: when there are lots of olap scanners to run for example 10000
+    // the rowsets maybe compacted when the last olap scanner starts
+    _tablet->obtain_header_rdlock();
+    OLAPStatus acquire_reader_st = _tablet->capture_rs_readers(_params.version, &_params.rs_readers);
+    _tablet->release_header_lock();
```

Comment:
I think you should add a function in tablet who will encapsulate the lock and unlock operation.

Reply:
currently could not. because some times we do step 1 step 2 step 3 and step 3 is capture_rs_readers.  If add lock in capture rs readers will be dead lock.